---
title: "Getting started with the scRNAutils scRNA pipeline"
author: "Jose Verdezoto Mosquera"
date:  "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{1. Getting started with the scRNAutils scRNA pipeline}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r echo=FALSE}
# These settings make the vignette prettier
knitr::opts_chunk$set(results="hold", collapse = FALSE, message=FALSE, 
                      warning = FALSE)
```

# Introduction to the scRNAutils pipeline 

The main goal of this tutorial is to show usage of the scRNA processing pipeline included in the `scRNAutils` package. The package is in a very early stage and will definitely undergo lots of improvements. 


## Installing scRNAutils and required dependencies 

Here are some important dependencies that need to be installed for the package to work:

```{r, eval=FALSE}

# Make sure the Bioconductor package manager is installed. 
# If not, install with the following line of code
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("Biocmanager")

# This function will check whether dependencies are installed. 
# If they're not, it'll install them either from CRAN or Bioconductor.
check.deps = function(p, bioc=FALSE) {
  if (! p %in% installed.packages()) {
    if (bioc) { 
      BiocManager::install(pkgs = p)
    } else {
      install.packages(p)
    }
  }
}

# Run the function to confirm installation of essential packages
check.deps("Seurat")
check.deps("scDblFinder", bioc=TRUE)
check.deps("SingleCellExperiment", bioc=TRUE)
check.deps("celda", bioc=TRUE)
check.deps("tidyverse")
check.deps("cluster")
check.deps("stats")
check.deps("data.table")
check.deps("ggsci")

```

Now we can install the `scRNAutils` package:

In the long term, this is probably how we'll set the installation

```{r, eval=FALSE}

# Use devtools for installation directly from Github
devtools::install_github("MillerLab-CPHG/scRNAutils")

```

Alternatively, since the repository is private for now, try this workflow instead:

  1. Clone repo:

```{bash, eval=FALSE}

# Clone repo
git clone git@github.com:MillerLab-CPHG/scRNAutils

```

  
  2. Build the package with `devtools`:

```{r, eval=FALSE}

# The following will create a file named scRNAutils_0.0.1.tar.gz
devtools::build("/Path/to/scRNAutils")

```

  
  3. I'm having some issues with `devtools::install()` but for now try this from the command line:

```{bash, eval=FALSE}

# Install the package 
R CMD INSTALL path/to/scRNAutils_0.0.1.tar.gz

```


Load all of the required packages into the R session:

```{r, message=FALSE, results="hide", warning=FALSE}

# Load all packages at once
deps = c("Seurat", "scDblFinder", "SingleCellExperiment", "celda", "tidyverse", "cluster", "broom", "stats", "data.table", "ggsci")
lapply(deps, library, character.only=TRUE)

```

## Overview of the pipeline and loading of example data

The pipeline, as currently written, is a single function that takes care of the QC, normalization, dimensionality reduction and clustering. Broadly this is the implemented workflow:

  1. Create Seurat object and do one round of clustering to get cell partitions needed for artificial doublet generation. 
  2. Remove doublets by performing n number of doublet identification iterations (n defaults to 3 but can be any number defined by the user).
  3. Once doublets have been removed, we use a `decontX` wrapper included in the package to remove ambient RNA. This wrapper takes a Seurat object as input and returns a Seurat object with cleaned raw counts within the RNA assay.
  4. Now the user can provide thresholds for number of UMIs, unique genes expressed and percentage of reads mapped to mithochondrial and hemoglobin genes. This will be the last stage of QC. 
  5. After QC, data will be normalized using `SCTransform`.
  6. Dimensionality reduction is done with PCA and UMAP.
  7. Now clusters can be identified. This step uses the user-provided range of clustering resolutions and calculate average silhouette scores for each of them.
  8. We then select the resolution with the maximum average silhouette score and perform Louvain clustering to get the final clusters.  

It's important to note that the pipeline currently takes a count matrix as input. The count matrix needs to be in sparse format. This is required by the user but we might eventually handle conversion to sparse matrix format indide of the function. 

```{r}

# Load the package
library(scRNAutils)

# Load example data, kindly provided by Paul Cheng at Stanford
cheng_rca_seurat = read_rds("/project/cphg-millerlab/Jose/human_scRNA_meta_analysis/scRNA_raw_matrices/Cheng_human_RCA_scRNA_data/RCA.rds")

# Let's split this dataset into 6 individual libraries
cheng_rca_seurat_list = Seurat::SplitObject(cheng_rca_seurat, 
                                            split.by = "Pt")

# We need to get the raw count matrices from each patient
extract_raw_counts = function(seurat_obj) { 
  return(seurat_obj@assays$RNA@counts)
}
raw_matrices_list = lapply(cheng_rca_seurat_list, 
                           extract_raw_counts)

# Now we select one of the matrices to test the pipeline
cheng_test_library = raw_matrices_list[[1]]
paste("The provided input is a", class(cheng_test_library))


```

# Run the pipeline and explore its outputs

## Process data with `doitall()`

The entire pipeline can be run using the `doitall()` function. We just need to specify some arguments. Description of the arguments can be found upon loading the package. But here's a brief overview:

   - `counts_matrix`: A sparse count matrix where rows are genes and columns are cells **(required)**. 
   - `study_id` and `library_id`: These are character strings that will add corresponding metadata. These are **required** metadata values. 
   - `min_res` and `max_res`: These are numeric values indicating the lower and upper boundaries of clustering resolutions for calculatin silhouette coefficients **(default=0.3 and 1.9)**. 
   - `dbl_remove_iter`: A numeric value defining the number of iterations performed to get consensus doublet barcodes **(default=3)**.
   - `genes_of_interest`: A character vector with the names of genes to check their expression before and after QC **(optional)**. 

```{r, message=FALSE, results="hide", warning=FALSE}

# Run pipeline
cheng_pipeline_outs = scRNAutils::doitall(counts_matrix = cheng_test_library,
                              study_id = "Cheng_et_al", 
                              library_id = "cheng_library_1", 
                              min_res = 0.3, 
                              max_res = 1, 
                              dbl_remove_iter = 3, 
                              genes_of_interest = c("MYH11", "LMOD1", "LTBP1", 
                                                    "CRTAC1", "IGFBP3", "FN1"))


```

The pipeline output consists of a list with 6 items:

  - The processed Seurat object.
  - Consensus doublet cell barcodes identified by n iterations defined by the user. 
  - UMAP of clusters before QC.
  - UMAP of clusters after QC.
  - Gene expression UMAPs before QC, including genes defined by the user in the genes_of_interest argument.
  - Gene expression UMAPs after QC, including genes defined by the user in the genes_of_interest argument.

```{r}

# Check items produced by the pipeline
names(cheng_pipeline_outs)

```

## Explore clustering results
  
We can explore clustering results as follows:

```{r, fig.cap="Clusters before and after QC", fig.align="center" }

# Plot cluster UMAPs comparing data before and after QC
before_qc_clusters = cheng_pipeline_outs$before_qc_clusters + ggtitle ("Clusters before QC") + custom_theme() + miller_discrete_scale()

after_qc_clusters = cheng_pipeline_outs$after_qc_clusters + ggtitle("Clusters after QC") + custom_theme() + miller_discrete_scale()

# Plot UMAPS together
cowplot::plot_grid(before_qc_clusters, after_qc_clusters)

```

## Explore gene expression results

We can also take a look at how gene expression looks before removing ambient RNA:

```{r, fig.cap="Gene expression for user defined genes before QC", fig.align="center"}

# Extract gene expression UMAPs from the pipeline output
before_qc_feature_list = cheng_pipeline_outs$before_qc_features 
after_qc_feature_list = cheng_pipeline_outs$after_qc_features 

# Plot gene expression before QC
cowplot::plot_grid(plotlist = before_qc_feature_list)

```

Now let's take a look at how gene expression looks after removing ambient RNA: 

```{r, fig.cap="Gene expression for user defined genes before QC", fig.align="center"}

cowplot::plot_grid(plotlist = after_qc_feature_list)

```

# Conclusion

That's it for now. As previously mentioned, the pipeline is in a very early stage but as of now, it does what I think are the essential processing steps for any scRNA library. Stay tuned for additional features and improvements.  

